src/pal/src/synchobj/mutex.cpp

```diff
@@ -613,3 +775,777 @@ DWORD SPINLOCKTryAcquire (LONG * lock)
+    :
+#if !HAVE_FULLY_FEATURED_PTHREAD_MUTEXES
+    m_timedWaiterCount(0),
+#endif // !HAVE_FULLY_FEATURED_PTHREAD_MUTEXES
+    m_lockOwnerProcessId(SharedMemoryHelpers::InvalidPid),
+    m_lockOwnerThreadId(SharedMemoryHelpers::InvalidPid),
+    m_isAbandoned(false)
+{
+#if !HAVE_FULLY_FEATURED_PTHREAD_MUTEXES
+    static_assert_no_msg(sizeof(m_timedWaiterCount) == sizeof(LONG)); // for interlocked operations
+#endif // HAVE_FULLY_FEATURED_PTHREAD_MUTEXES
+
+    _ASSERTE(SharedMemoryManager::IsCreationDeletionProcessLockAcquired());
+    _ASSERTE(SharedMemoryManager::IsCreationDeletionFileLockAcquired());
+
+#if HAVE_FULLY_FEATURED_PTHREAD_MUTEXES
+    MutexHelpers::InitializeProcessSharedRobustRecursiveMutex(&m_lock);
+#endif // HAVE_FULLY_FEATURED_PTHREAD_MUTEXES
+}
+
+NamedMutexSharedData::~NamedMutexSharedData()
+{
+    _ASSERTE(SharedMemoryManager::IsCreationDeletionProcessLockAcquired());
+    _ASSERTE(SharedMemoryManager::IsCreationDeletionFileLockAcquired());
+
+#if HAVE_FULLY_FEATURED_PTHREAD_MUTEXES
+    MutexHelpers::DestroyMutex(&m_lock);
+#endif // HAVE_FULLY_FEATURED_PTHREAD_MUTEXES
+}
+
+#if HAVE_FULLY_FEATURED_PTHREAD_MUTEXES
+pthread_mutex_t *NamedMutexSharedData::GetLock()
+{
+    return &m_lock;
+}
+#else // !HAVE_FULLY_FEATURED_PTHREAD_MUTEXES
+bool NamedMutexSharedData::HasAnyTimedWaiters() const
+{
+    return
+        InterlockedCompareExchange(
+            const_cast<LONG *>(reinterpret_cast<const LONG *>(&m_timedWaiterCount)),
+            -1 /* Exchange */,
+            -1 /* Comparand */) != 0;
+}
+
+void NamedMutexSharedData::IncTimedWaiterCount()
+{
+    ULONG newValue = InterlockedIncrement(reinterpret_cast<LONG *>(&m_timedWaiterCount));
+    if (newValue == 0)
+    {
+        throw SharedMemoryException(static_cast<DWORD>(SharedMemoryError::OutOfMemory));
+    }
+}
+
+void NamedMutexSharedData::DecTimedWaiterCount()
+{
+    ULONG newValue = InterlockedDecrement(reinterpret_cast<LONG *>(&m_timedWaiterCount));
+    _ASSERTE(newValue + 1 != 0);
+}
+#endif // HAVE_FULLY_FEATURED_PTHREAD_MUTEXES
+
+bool NamedMutexSharedData::IsAbandoned() const
+{
+    _ASSERTE(IsLockOwnedByCurrentThread());
+    return m_isAbandoned;
+}
+
+void NamedMutexSharedData::SetIsAbandoned(bool isAbandoned)
+{
+    _ASSERTE(IsLockOwnedByCurrentThread());
+    _ASSERTE(m_isAbandoned != isAbandoned);
+
+    m_isAbandoned = isAbandoned;
+}
+
+bool NamedMutexSharedData::IsLockOwnedByAnyThread() const
+{
+    return m_lockOwnerProcessId != SharedMemoryHelpers::InvalidPid || m_lockOwnerThreadId != SharedMemoryHelpers::InvalidPid;
+}
+
+bool NamedMutexSharedData::IsLockOwnedByCurrentThread() const
+{
+    return m_lockOwnerProcessId == GetCurrentProcessId() && m_lockOwnerThreadId == GetCurrentThreadId();
+}
+
+void NamedMutexSharedData::SetLockOwnerToCurrentThread()
+{
+    m_lockOwnerProcessId = GetCurrentProcessId();
+    _ASSERTE(m_lockOwnerProcessId != SharedMemoryHelpers::InvalidPid);
+    m_lockOwnerThreadId = GetCurrentThreadId();
```

Comment:
I see, and it looks like gettid() is the wrong thing as well. Fixed to use the function you suggested, and changed all thread IDs stored in shared-memory to 64-bit ints.